# Optimisation 4. Going back to the raw data, and keeping column NAME.

## Preprocessing

In [1]:
# Import our dependencies
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
import pandas as pd
import tensorflow as tf
import tensorflow.keras as keras

#  Import and read the charity_data.csv.
import pandas as pd 
application_df = pd.read_csv("../Resources/charity_data.csv")
application_df.head()

EIN                                      NAME APPLICATION_TYPE  \
0  10520599              BLUE KNIGHTS MOTORCYCLE CLUB              T10   
1  10531628    AMERICAN CHESAPEAKE CLUB CHARITABLE TR               T3   
2  10547893        ST CLOUD PROFESSIONAL FIREFIGHTERS               T5   
3  10553066            SOUTHSIDE ATHLETIC ASSOCIATION               T3   
4  10556103  GENETIC RESEARCH INSTITUTE OF THE DESERT               T3   

        AFFILIATION CLASSIFICATION      USE_CASE  ORGANIZATION  STATUS  \
0       Independent          C1000    ProductDev   Association       1   
1       Independent          C2000  Preservation  Co-operative       1   
2  CompanySponsored          C3000    ProductDev   Association       1   
3  CompanySponsored          C2000  Preservation         Trust       1   
4       Independent          C1000     Heathcare         Trust       1   

      INCOME_AMT SPECIAL_CONSIDERATIONS  ASK_AMT  IS_SUCCESSFUL  
0              0                      N     5000              1  
1         1-9999                      N   108590              1  
2              0                      N     5000              0  
3    10000-24999                      N     6692              1  
4  100000-499999                      N   142590              1

In [2]:
# Drop the non-beneficial ID column'EIN'
application_df = application_df.drop(['EIN'], axis=1)
application_df

NAME APPLICATION_TYPE  \
0                           BLUE KNIGHTS MOTORCYCLE CLUB              T10   
1                 AMERICAN CHESAPEAKE CLUB CHARITABLE TR               T3   
2                     ST CLOUD PROFESSIONAL FIREFIGHTERS               T5   
3                         SOUTHSIDE ATHLETIC ASSOCIATION               T3   
4               GENETIC RESEARCH INSTITUTE OF THE DESERT               T3   
...                                                  ...              ...   
34294              THE LIONS CLUB OF HONOLULU KAMEHAMEHA               T4   
34295           INTERNATIONAL ASSOCIATION OF LIONS CLUBS               T4   
34296                                PTA HAWAII CONGRESS               T3   
34297  AMERICAN FEDERATION OF GOVERNMENT EMPLOYEES LO...               T5   
34298                           WATERHOUSE CHARITABLE TR               T3   

            AFFILIATION CLASSIFICATION      USE_CASE  ORGANIZATION  STATUS  \
0           Independent          C1000    ProductDev   Association       1   
1           Independent          C2000  Preservation  Co-operative       1   
2      CompanySponsored          C3000    ProductDev   Association       1   
3      CompanySponsored          C2000  Preservation         Trust       1   
4           Independent          C1000     Heathcare         Trust       1   
...                 ...            ...           ...           ...     ...   
34294       Independent          C1000    ProductDev   Association       1   
34295  CompanySponsored          C3000    ProductDev   Association       1   
34296  CompanySponsored          C2000  Preservation   Association       1   
34297       Independent          C3000    ProductDev   Association       1   
34298       Independent          C1000  Preservation  Co-operative       1   

          INCOME_AMT SPECIAL_CONSIDERATIONS   ASK_AMT  IS_SUCCESSFUL  
0                  0                      N      5000              1  
1             1-9999                      N    108590              1  
2                  0                      N      5000              0  
3        10000-24999                      N      6692              1  
4      100000-499999                      N    142590              1  
...              ...                    ...       ...            ...  
34294              0                      N      5000              0  
34295              0                      N      5000              0  
34296              0                      N      5000              0  
34297              0                      N      5000              1  
34298          1M-5M                      N  36500179              0  

[34299 rows x 11 columns]

In [3]:
# Look at APPLICATION_TYPE value counts for droping unwanted values
app_counts = application_df['APPLICATION_TYPE'].value_counts()
app_counts

T3     27037
T4      1542
T6      1216
T5      1173
T19     1065
T8       737
T7       725
T10      528
T9       156
T13       66
T12       27
T2        16
T25        3
T14        3
T29        2
T15        2
T17        1
Name: APPLICATION_TYPE, dtype: int64

In [4]:
# Choose a cutoff value and create a list of application types to be replaced
# use the variable name `application_types_to_replace`

application_types_to_replace = list(app_counts[app_counts<500].index) 

# Replace in dataframe
for app in application_types_to_replace:
    application_df['APPLICATION_TYPE'] = application_df['APPLICATION_TYPE'].replace(app,"Other")

# Check to make sure binning was successful
application_df['APPLICATION_TYPE'].value_counts()

T3       27037
T4        1542
T6        1216
T5        1173
T19       1065
T8         737
T7         725
T10        528
Other      276
Name: APPLICATION_TYPE, dtype: int64

In [5]:
# Look at CLASSIFICATION value counts for binning
class_counts = application_df['CLASSIFICATION'].value_counts()
class_counts

C1000    17326
C2000     6074
C1200     4837
C3000     1918
C2100     1883
         ...  
C4120        1
C8210        1
C2561        1
C4500        1
C2150        1
Name: CLASSIFICATION, Length: 71, dtype: int64

In [6]:
# Choose a cutoff value and create a list of classifications to be replaced
# use the variable name `classifications_to_replace`

classifications_to_replace = list(class_counts[class_counts<1000].index) 

# Replace in dataframe
for cls in classifications_to_replace:
    application_df['CLASSIFICATION'] = application_df['CLASSIFICATION'].replace(cls,"Other")
    
# Check to make sure binning was successful
application_df['CLASSIFICATION'].value_counts()

C1000    17326
C2000     6074
C1200     4837
Other     2261
C3000     1918
C2100     1883
Name: CLASSIFICATION, dtype: int64

In [7]:
# Finding data types in the DataFrame
types_list = application_df.dtypes
types_list

NAME                      object
APPLICATION_TYPE          object
AFFILIATION               object
CLASSIFICATION            object
USE_CASE                  object
ORGANIZATION              object
STATUS                     int64
INCOME_AMT                object
SPECIAL_CONSIDERATIONS    object
ASK_AMT                    int64
IS_SUCCESSFUL              int64
dtype: object

In [8]:
for item in types_list.index:
    if types_list[item] == 'int64':
        item
        types_list = types_list.drop(item)
types_list

NAME                      object
APPLICATION_TYPE          object
AFFILIATION               object
CLASSIFICATION            object
USE_CASE                  object
ORGANIZATION              object
INCOME_AMT                object
SPECIAL_CONSIDERATIONS    object
dtype: object

In [9]:
# Convert categorical data to numeric with `pd.get_dummies`
application_df = pd.get_dummies(application_df, columns=types_list.index)
application_df

STATUS   ASK_AMT  IS_SUCCESSFUL  \
0           1      5000              1   
1           1    108590              1   
2           1      5000              0   
3           1      6692              1   
4           1    142590              1   
...       ...       ...            ...   
34294       1      5000              0   
34295       1      5000              0   
34296       1      5000              0   
34297       1      5000              1   
34298       1  36500179              0   

       NAME_1 DAY RANCH RESCUE AND RURAL OKLAHOMA ANIMAL RESOURCE INC  \
0                                                      0                
1                                                      0                
2                                                      0                
3                                                      0                
4                                                      0                
...                                                  ...                
34294                                                  0                
34295                                                  0                
34296                                                  0                
34297                                                  0                
34298                                                  0                

       NAME_100 BLACK MEN OF AMERICA  NAME_100 BLACK MEN OF MEMPHIS INC  \
0                                  0                                  0   
1                                  0                                  0   
2                                  0                                  0   
3                                  0                                  0   
4                                  0                                  0   
...                              ...                                ...   
34294                              0                                  0   
34295                              0                                  0   
34296                              0                                  0   
34297                              0                                  0   
34298                              0                                  0   

       NAME_100 BLACK MEN OF WEST GEORGIA INC  NAME_1150 WEBSTER STREET INC  \
0                                           0                             0   
1                                           0                             0   
2                                           0                             0   
3                                           0                             0   
4                                           0                             0   
...                                       ...                           ...   
34294                                       0                             0   
34295                                       0                             0   
34296                                       0                             0   
34297                                       0                             0   
34298                                       0                             0   

       NAME_116TH CAVALRY REGIMENT CHAPTER OF THE US CAVALRY & ARMOR ASSOCIATION  \
0                                                      0                           
1                                                      0                           
2                                                      0                           
3                                                      0                           
4                                                      0                           
...                                                  ...                           
34294                                                  0                           
34295                                                  0                           
34296              

In [10]:
# Split our preprocessed data into our features and target arrays
X = application_df.drop('IS_SUCCESSFUL', axis=1).values
y = application_df['IS_SUCCESSFUL'].values


# Split the preprocessed data into a training and testing dataset
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state = 42, stratify=y)

In [11]:
# Create a StandardScaler instances
scaler = StandardScaler()

# Fit the StandardScaler
X_scaler = scaler.fit(X_train)

# Scale the data
X_train_scaled = X_scaler.transform(X_train)
X_test_scaled = X_scaler.transform(X_test)

In [12]:
input_dimensions = X_train_scaled.shape[1]
input_dimensions

19611

# Compile, Train and Evaluate the Model

In [13]:
#Letting keras-tuner decide activation function, number of layers and number of neurons

In [14]:
# Create a method that creates a new Sequential model with hyperparameter options
def create_model(hp):
    nn_model = tf.keras.models.Sequential()

    # Allow kerastuner to decide which activation function to use in hidden layers
    activation = hp.Choice('activation',['relu','tanh','sigmoid'])
    
    # Allow kerastuner to decide number of neurons in first layer
    nn_model.add(tf.keras.layers.Dense(units=hp.Int('first_units',
        min_value=1,
        max_value=80,
        step=5), activation=activation, input_dim=input_dimensions))

    # Allow kerastuner to decide number of hidden layers and neurons in hidden layers
    for i in range(hp.Int('num_layers', 1, 6)):
        nn_model.add(tf.keras.layers.Dense(units=hp.Int('units_' + str(i),
            min_value=1,
            max_value=100,
            step=5),
            activation=activation))
    
    nn_model.add(tf.keras.layers.Dense(units=1, activation="sigmoid"))

    # Compile the model
    nn_model.compile(loss="binary_crossentropy", optimizer='adam', metrics=["accuracy"])
    
    return nn_model

In [15]:
# Import the kerastuner library
import keras_tuner as kt

tuner = kt.Hyperband(
    create_model,
    objective="val_accuracy",
    overwrite=True,
    max_epochs=20,
    hyperband_iterations=2,)


In [16]:
# Run the kerastuner search for best hyperparameters
tuner.search(X_train_scaled,y_train,epochs=20,validation_data=(X_test_scaled,y_test))

Trial 60 Complete [00h 09m 58s]
val_accuracy: 0.7870553731918335

Best val_accuracy So Far: 0.7953352928161621
Total elapsed time: 03h 31m 05s
INFO:tensorflow:Oracle triggered exit


In [17]:
# Get best model hyperparameters
best_hyper = tuner.get_best_hyperparameters(1)[0]
best_hyper.values

{'activation': 'sigmoid',
 'first_units': 46,
 'num_layers': 4,
 'units_0': 76,
 'units_1': 96,
 'units_2': 31,
 'units_3': 96,
 'units_4': 76,
 'units_5': 31,
 'tuner/epochs': 3,
 'tuner/initial_epoch': 0,
 'tuner/bracket': 2,
 'tuner/round': 0}

In [18]:
# Evaluate best model against full test data
best_model = tuner.get_best_models(1)[0]
model_loss, model_accuracy = best_model.evaluate(X_test_scaled,y_test,verbose=2)
print(f"Loss: {model_loss}, Accuracy: {model_accuracy}")

268/268 - 4s - loss: 0.4520 - accuracy: 0.7953 - 4s/epoch - 14ms/step
Loss: 0.45202428102493286, Accuracy: 0.7953352928161621


In [19]:
# Export our model to HDF5 file
best_model.save('../HDF5_files/AlphabetSoupCharity_Optimization_keep_NAME.h5')

## The model has achieved 79.5% accuracy, more than the 75% target